service/history/workflow/delete_manager.go

```diff
@@ -0,0 +1,248 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination delete_manager_mock.go
+
+package workflow
+
+import (
+	"context"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/searchattribute"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/worker/archiver"
+)
+
+type (
+	DeleteManager interface {
+		DeleteDeletedWorkflowExecution(namespaceID namespace.ID, we commonpb.WorkflowExecution, weCtx Context, ms MutableState, transferTaskVersion int64) error
+		DeleteWorkflowExecutionRetention(namespaceID namespace.ID, we commonpb.WorkflowExecution, weCtx Context, ms MutableState, timerTaskVersion int64) error
```

Comment:
DeleteDeleted? What does that mean? Could you put some comment on them, what are the differences of these 2 APIs.
And, ctx is usually first input argument.

Reply:
+1. DeleteDeleted will be used by transfer queue I guess? 
nit : `DeleteForWorkflowRetention` makes more sense to me but no strong opinion.

Also, probably do not reference any queue type (transfer, time) in new code as we are doing the refactoring for injecting private queue. Maybe `taskVersion` is sufficient in this case.

Reply:
+1. DeleteDeleted will be used by transfer queue I guess?
nit : DeleteForWorkflowRetention makes more sense to me but no strong opinion.

Also, probably do not reference any queue type (transfer, time) in new code as we are doing the refactoring for queue processors. Maybe taskVersion is sufficient in this case.

Reply:
This PR was cherry-picked from another bigger PR. `DeleteDeleted` makes perfect sense there but here it looks confusing, I agree. I am renaming it. At least for now.
This is `workflow.Context` not `context.Context`, so I believe it is ok it pass it with other workflow related params.
Renaming `DeleteWorkflowExecutionRetention` to `DeleteWorkflowExecutionByRetention`.
And also giving task version parameters more generic names.